public-api/src/main/java/org/pmiops/workbench/google/GoogleAnalyticsServiceImpl.java

```diff
@@ -0,0 +1,108 @@
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import javax.inject.Provider;
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLEncoder;
+import java.nio.charset.StandardCharsets;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+@Service
+public class GoogleAnalyticsServiceImpl implements GoogleAnalyticsService {
+
+    private final HTTPHeader header =
+            new HTTPHeader("Content-Type", "application/x-www-form-urlencoded");
+    private URLFetchService urlFetchService;
+    private final Provider<WorkbenchConfig> configProvider;
+
+    @Autowired
+    public GoogleAnalyticsServiceImpl(Provider<WorkbenchConfig> configProvider) {
+        this.configProvider = configProvider;
+        setUrlFetchService(getUrlFetchService());
+    }
+
+    /**
+     * Posts an Event Tracking message to Google Analytics.
+     * See https://developers.google.com/analytics/devguides/collection/gtagjs/events
+     * for a list of default GA events.
+     *
+     * @param category the required event category, e.g. 'engagement'
+     * @param action the required event action, e.g. 'search'
+     * @param label the optional event label, e.g. 'search_term'
+     * @param value the optional value, e.g. 'diabetes'
+     * @return int value of the http response.
+     * @exception IOException if the URL could not be posted to
+     */
+    public int trackEventToGoogleAnalytics(
+            String category, String action, String label, String value) throws IOException {
+        Map<String, String> map = new LinkedHashMap<>();
+        map.put("v", "1");             // Version.
+        map.put("tid", configProvider.get().server.gaId);
+        map.put("cid", configProvider.get().server.clientId);
+        map.put("t", "event");         // Event hit type.
+        map.put("ec", encode(category, true));
+        map.put("ea", encode(action, true));
+        map.put("el", encode(label, false));
+        map.put("ev", encode(value, false));
+
+        HTTPRequest request = new HTTPRequest(getGoogleAnalyticsEndpoint(), HTTPMethod.POST);
+        request.addHeader(this.header);
+        request.setPayload(getPostData(map));
+
+        HTTPResponse httpResponse = urlFetchService.fetch(request);
+        return httpResponse.getResponseCode();
+    }
+
+    @VisibleForTesting
+    public URL getGoogleAnalyticsEndpoint() throws MalformedURLException {
+        return new URL("http", "www.google-analytics.com", "/collect");
+    }
+
+    @VisibleForTesting
+    public void setUrlFetchService(URLFetchService fetchService) {
+        this.urlFetchService = fetchService;
+    }
+
+    private URLFetchService getUrlFetchService() {
+        return URLFetchServiceFactory.getURLFetchService();
+    }
+
+    private static byte[] getPostData(Map<String, String> map) {
```

Comment:
This format seems pretty custom - where did you find this documented?

Reply:
It is. This comes from Google's example implementation: https://cloud.google.com/appengine/docs/standard/java/google-analytics

Reply:
Aha , thanks.